Fixes #3650: Support stop-all-threads mode debugging for multi-thread…

edumunoz opened the pull request:
As discussed in #3650, this adds support for stop-all-threads mode by adding a flag to the StoppedEvent that the debug adapter sends. Here's the fix in action:

<img src="https://cloud.githubusercontent.com/assets/637952/13693056/dd2e406c-e6fb-11e5-98b3-eeb0390d3794.gif" data-canonical-src="https://cloud.githubusercontent.com/assets/637952/13693056/dd2e406c-e6fb-11e5-98b3-eeb0390d3794.gif" width="335" height="315" />

I tested this change with a basic multi-threaded program, and verified:
- [x] Expanding threads that are not the one that sent the stopped event triggers a query for their callstack
- [x] Already expanded threads get updated when a new stopped event is sent
- [x] Node debugging works (including debugging VSCode extensions)
- [x] Mono debugging works (handling multiple threads work as before)
- [x] Go debugging works (handling multiple threads work as before)
- [x] C# debugging works (with the new omnisharp-based extension)

msftclas commented:
Hi __@edumunoz__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

edumunoz commented:
I changed my username on Github after linking my account with my corporate account. Hence it's asking for CLA. Anyway of re-triggering msftclas?

msftclas commented:
__@edumunoz__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

weinand commented:
From a debug protocol perspective I suggest to rename the `stopAllThreads` attribute on the StoppedEvent to `allThreadsAreStopped`. `stopAllThreads` sounds like an action, whereas `allThreadsAreStopped` makes clear that this is additional information on the StoppedEvent.

@isidorn could you please review the PR from the front-end perspective.

isidorn commented:
@edumunoz thanks for the PR.
This code change impacts all debugging scenarios, so most of the debug adapters need to be tested as this can easily break lots of user scenarios - I have added checkboxes to your original comment.
Unfortunetely, we do not have tests for the debug service.

Concerns which I have regarding how model data is handled I will comment directly in the commits.

felixfbecker commented:
I don't understand the change of this PR. Could someone please explain?

isidorn commented:
@felixfbecker #3650 has more details

edumunoz commented:
Thanks for the feedback, @isidorn!

I'll work through it and move the logic to `Thread`. I left a couple of replies/comments.

edumunoz commented:
@isidorn, I updated the PR addressing your feedback, as well as adding some testing.

For testing the existing debug adapters, on top of what is in brackets (which is the stuff that needs to be tested thoroughly), is there a standard list of stuff to test (e.g., list of stuff for breakpoints, list of stuff for step, etc.)?

isidorn commented:
@edumunoz great! It looks much better now. I will add some comments directly in the code.

For testing, I would focus on the things in brackets, a smoke test of 'node' debugging with some focus on clicking in the call stack should do the trick. For 'go' and 'mono' focussing on threads makes most sense since changes here refactor how we deal with threads.
Other debuggers like powershell and chrome should not use additional threads, and just checking 'node' should cover their case. @daviwil can you please confirm if powershell uses multiple threads?

I guess that you can already build your custom version of code. For testing 'mono' and 'go' you would need to install those extension in your custom built vscode, or just copy over those extensions to your userFolder\\.vscode-oss\extensions 

Here's a part of our smoke test that covers debugging, so doing something like this while focussing on the call stack would be best for 'node'. If you do not have access to out smoketest-express, taking any node app out there should do the trick. E.g https://github.com/bpasero/standup

#### Debugging JavaScript

	• git clone https://monacotools.visualstudio.com/DefaultCollection/Samples/_git/vscode-smoketest-express
        •  in VSCode switch to debugging viewlet
	• Click on the gear to create a default launch.json file:
		make sure that VSCode automatically detects ./bin/www as the 'program' attribute
	• Select Launch in the dropdown at the top of the debug view
		verify that the drop-down menu shows entries from launch.json
	• set a breakpoint in index.js:6
	• press F5 to start debugging. Verify:
		workbench transforms into "debug mode" - glyph margin and status bar turns orange
	• open browser at http://localhost:3000/ 
		verify the breakpoint in index.js gets hit
	• Verify step over, step in, continue work
	• while stopped, verify:
		○ you can hover over values
		○ clicking on a stack frame shows variables associated with that stack frame
		○ clicking on a breakpoint takes you to that breakpoint
		○ you can inspect values of complex variables
		○ you can add a watch expression
		○ You can evaluate a simple expression in the debug console
	• Verify you can press Restart and the server restarts and you are back in debug mode
       • Verify you can stop the debug session and that workbench exits "debug mode"

daviwil commented:
@isidorn For now we don't deal with multithreading, not sure if it's something we'll have to support.  Thanks for the notification!

cc @rkeithhill

edumunoz commented on src/vs/workbench/parts/debug/common/debug.ts:
```diff
@@ -55,23 +55,30 @@ export interface IThread extends ITreeElement {
+	 * Indicates whether this thread is stopped. The callstack for stopped
+	 * threads can be retrieved from the debug adapter.
+	 */
+	stopped: boolean;
```
`stopped` doesn't necessarily need to be exposed. We could have two methods `setRunning` and `setStopped` that would set `stopped` to false and `true`, respectively.

I left it exposed so that the tests can check it. I'm fine removing `stopped` from the interface and only have setters. Let me know what you think.

isidorn replied:
Having `stopped` exposed like this is fine I think. It is a general property that other users of `Thread` might be interested in the future. So I would leave it as it is.

edumunoz commented on src/vs/workbench/parts/debug/common/debugModel.ts:
```diff
@@ -414,6 +424,14 @@ export class Model extends ee.EventEmitter implements debug.IModel {
 		this.emit(debug.ModelEvents.CALLSTACK_UPDATED);
 	}
 
+	public continueThreads(): void {
```
I could add the functionality of `continueThreads` to `clearThreads` behind yet another boolean arguments.

I think having an explicit method is better (`clearThreads` is already confusing enough). Let me know what you think.

isidorn replied:
Yes, `clearThreads` is already confusing enough. Explicit `continueThreads` is fine imho.

edumunoz commented:
@isidorn, thanks for your thorough review. In the end I went for exposing a non-promise based `callstack` in the `Thread` model, that way the changes to other parts of the codebase are limited to renaming the `callstack` property.

I left a couple of comments in 2 places to get your opinion.

It would be great if you could give us the heads up today that the code looks good (bar _very small_ changes), so we can spend more time testing the partner debug adapters thoroughly.

isidorn commented:
Ok, I'll review your newest changes today - and I will also comment on your older questions even though we decided to take the non-promise based `callStack` in the `Thread` model. Which I think is a good decision for now to make the spread of changes minimal.

Squashing the commits would be great!

isidorn commented on src/vs/workbench/parts/debug/browser/debugEditorModelManager.ts:
```diff
@@ -134,18 +134,18 @@ export class DebugEditorModelManager implements IWorkbenchContribution {
 		const result: editorcommon.IModelDeltaDecoration[] = [];
```
I like that there are not a lot of changes in this file now. Due to this we will not have to intensively test the UI changes in the editor (e.g. breakpoints).

isidorn commented on src/vs/workbench/parts/debug/browser/debugViewer.ts:
```diff
@@ -189,6 +189,10 @@ export class BaseDebugController extends treedefaults.DefaultController {
 
```
Looks good.

isidorn commented on src/vs/workbench/parts/debug/browser/debugViewlet.ts:
```diff
@@ -93,7 +93,7 @@ class VariablesView extends viewlet.CollapsibleViewletView {
 		this.toDispose.push(this.debugService.addListener2(debug.ServiceEvents.STATE_CHANGED, () => {
```
Looks good.

isidorn commented on src/vs/workbench/parts/debug/common/debugModel.ts:
```diff
@@ -90,16 +91,61 @@ export function getFullExpressionName(expression: debug.IExpression, sessionType
+	public stopped: boolean;
 
-	constructor(public name: string, public threadId, public callStack: debug.IStackFrame[]) {
+	constructor(public name: string, public threadId) {
```
Not important - but I personally am not a big fun of using `undefined`. I always try to use `null`, because having one `null` is already bad enough. You do not have to change anything here, but just for the future so you know that is my style.

isidorn commented on src/vs/workbench/parts/debug/common/debugSource.ts:
```diff
@@ -40,8 +40,8 @@ export class Source {
 			// first try to find the raw source amongst the stack frames - since that represenation has more data (source reference),
```
Looks good.

isidorn commented on src/vs/workbench/parts/debug/test/common/debugModel.test.ts:
```diff
@@ -7,6 +7,11 @@ import assert = require('assert');
 import uri from 'vs/base/common/uri';
```
Really happy for the tests.

isidorn commented on src/vs/workbench/parts/debug/common/debug.ts:
```diff
@@ -25,6 +25,7 @@ export interface IRawModelUpdate {
 	thread?: DebugProtocol.Thread;
```
Looks good.

isidorn commented on src/vs/workbench/parts/debug/electron-browser/debugService.ts:
```diff
@@ -240,11 +240,16 @@ export class DebugService extends ee.EventEmitter implements debug.IDebugService
+					stoppedDetails: event.body,
+					allThreadsStopped: event.body.allThreadsStopped
+				});
+
```
Looks good - I like that the changes here are minimal now.

Though I think you can handle this promises a bit better - you can nicely have only one place where you handle errors. I propose:
* Line 242: needs a `done`
* Line 251: needs a `return` and a `then`
* Line 256: needs a `return` and no `done` - this will be boobled up
* Move error handling from line 260 to 261

This way I believe the error handling and `promise` flow is nicer.

edumunoz replied:
Makes sense.

isidorn commented on src/vs/workbench/parts/debug/common/debugModel.ts:
```diff
@@ -90,16 +91,61 @@ export function getFullExpressionName(expression: debug.IExpression, sessionType
+		if (!this.stopped) {
+			return TPromise.as([]);
+		}
+		if (!this.callStackQueried) {
```
I believe the `callStackQueried` flag is superflous and we can remove it. What I propose is to just change line 126 to `if (!this.promisedCallStack)`. This way if we already queried a call to the adapter the original call will be returned and no additional calls will be made. Same as the `ExpressionsContainer` does with `getChildren`

edumunoz replied:
Good catch. It should behave the same, just getting rid of the flag.

isidorn commented on src/vs/workbench/parts/debug/common/debugModel.ts:
```diff
@@ -378,6 +424,14 @@ export class Model extends ee.EventEmitter implements debug.IModel {
 		this.emit(debug.ModelEvents.CALLSTACK_UPDATED);
 	}
 
```
Should `continueThreads()` also `clearCallstack()` for each `thread`?

edumunoz replied:
I added a line to `clearThreads`.

isidorn commented on src/vs/workbench/parts/debug/common/debugModel.ts:
```diff
@@ -13,6 +13,7 @@ import severity from 'vs/base/common/severity';
 import types = require('vs/base/common/types');
```
Looks good overall - some comments below.

edumunoz commented on src/vs/workbench/parts/debug/common/debugModel.ts:
```diff
@@ -624,11 +676,13 @@ export class Model extends ee.EventEmitter implements debug.IModel {
 
 	public sourceIsUnavailable(source: Source): void {
 		Object.keys(this.threads).forEach(key => {
-			this.threads[key].callStack.forEach(stackFrame => {
```
For some reason I had missed updating this reference to `callStack`.

Just pointing out what we have changed today that wasn't in yesterday's commit and you hadn't commented on.

edumunoz commented:
@isidorn, I addressed your feedback from yesterday.
We performed the testing tasks for different platforms based on the items in `Debugging JavaScript` *(checked off the items in the description of this PR).

For the existing debuggers, we tried that the behavior remains identical.
For new debuggers that support `allThreadsStopped`, the multi-threading support kicks in and the experience is as we expected.

Let us know if you need anything else from us.

isidorn commented:
Looks good to me! Great work - merging!

THE-ORIGINAL-MASTER commented:
Well are you all loving the bugs
